pkg/networkpath/traceroute/tcp/utils.go

```diff
@@ -243,145 +243,83 @@ func handlePackets(ctx context.Context, conn rawConnWrapper, listener string, lo
 	}
 }
 
-// MarshalPacket takes in an ipv4 header and a payload and copies
-// them into a newly allocated []byte
-func MarshalPacket(header *ipv4.Header, payload []byte) ([]byte, error) {
-	hdrBytes, err := header.Marshal()
-	if err != nil {
-		return nil, err
-	}
-
-	packet := make([]byte, len(hdrBytes)+len(payload))
-	copy(packet[:len(hdrBytes)], hdrBytes)
-	copy(packet[len(hdrBytes):], payload)
-
-	return packet, nil
-}
-
-// readRawPacket creates a gopacket given a byte array
-// containing a packet
-//
-// TODO: try doing this manually to see if it's more performant
-// we should either always use gopacket or never use gopacket
-func readRawPacket(rawPacket []byte) gopacket.Packet {
-	return gopacket.NewPacket(rawPacket, layers.LayerTypeIPv4, gopacket.Default)
-}
-
-func parseIPv4Layer(pkt gopacket.Packet) (net.IP, net.IP, error) {
-	if ipLayer := pkt.Layer(layers.LayerTypeIPv4); ipLayer != nil {
-		ip, ok := ipLayer.(*layers.IPv4)
-		if !ok {
-			return net.IP{}, net.IP{}, fmt.Errorf("failed to assert IPv4 layer type")
-		}
-
-		return ip.SrcIP, ip.DstIP, nil
-	}
-
-	return net.IP{}, net.IP{}, fmt.Errorf("packet does not contain an IPv4 layer")
-}
-
-func parseICMP(header *ipv4.Header, payload []byte) (*icmpResponse, error) {
-	packetBytes, err := MarshalPacket(header, payload)
-	if err != nil {
-		return nil, fmt.Errorf("failed to marshal packet: %w", err)
-	}
-	packet := readRawPacket(packetBytes)
-
-	return parseICMPPacket(packet)
-}
-
-// parseICMPPacket takes in a gopacket.Packet and tries to convert to an ICMP message
-// it returns all the fields from the packet we need to validate it's the response
+// parseICMP takes in an IPv4 header and payload and tries to convert to an ICMP
+// message, it returns all the fields from the packet we need to validate it's the response
 // we're looking for
-func parseICMPPacket(pkt gopacket.Packet) (*icmpResponse, error) {
-	// this parsing could likely be improved to be more performant if we read from the
-	// the original packet bytes directly where we expect the required fields to be
-	// or even just creating a single DecodingLayerParser but in both cases we lose
-	// some flexibility
-	var err error
+func parseICMP(header *ipv4.Header, payload []byte) (*icmpResponse, error) {
+	// in addition to parsing, it is probably not a bad idea to do some validation
+	// so we can ignore the ICMP packets we don't care about
 	icmpResponse := icmpResponse{}
 
-	icmpResponse.SrcIP, icmpResponse.DstIP, err = parseIPv4Layer(pkt)
-	if err != nil {
+	if header.Protocol != 1 || header.Version != 4 ||
```

Comment:
Can we use constants instead of numbers, for readability/maintenability :) 

```suggestion
	if header.Protocol != unix.IPPROTO_ICMP || header.Version != ipv4.Version ||
```

Reply:
These won't compile on Windows.

For now I'm going to define a constant above. We still plan to support Windows in the future and I don't think we should add a build constraint just to import two constants